### ISSUE TITLE
Aggregates should materialise once for every unique aggregate

### ISSUE THREAD

**Issue by rdowavic** [souffle-lang/souffle#1433]
At the moment, even if two aggregates are equivalent (ie will result in the same value), they will materialise twice into two different relations, failing to recognise that this is redundant.

```
D(x) :- x < sum y : { C(y), B(y) }, B(x), A(z), z = sum y : { C(y), B(y) }.
```
becomes
```
__agg_rel_1(z0) :- 
   z0 = sum  y1 : { C( y1), B( y1) }.

__agg_rel_2(z1) :- 
   z1 = sum  y2 : { C( y2), B( y2) }.

D(x) :- 
   +disconnected0(),
   x < z0,
   B(x),
   __agg_rel_1(z0).

+disconnected0() :- 
   A(z),
   z = z1,
   __agg_rel_2(z1).
```
(disabling ResolveAliases and MaterializeAggregationQueries for clarity)

Both the disconnected0 and D predicates could have relied on the same materialised aggregate relations.

**Comment by azreika** [comment]
This also occurs with `PartitionBodyLiteralsTransformer`, where we might have two disconnected relations from two separate rules that are identical. E.g.:

```
 d(X) :- c(X,Y), a(Y), a(Z), b(Z).
 d(X) :- c(X,Y), a(X), a(Z), b(Z).
```

would produce something like:

```
disconnected0() :- a(Z), b(Z).
disconnected1() :- a(Z), b(Z).
 d(X) :- c(X,Y), a(Y), disconnected0().
 d(X) :- c(X,Y), a(X), disconnected1().
```
which should be reduced to:
```
disconnected0() :- a(Z), b(Z).
 d(X) :- c(X,Y), a(Y), disconnected0().
```

In my opinion, removing these is more the job of `MinimiseProgramTransformer` later on in the AST pipeline. It was already extended at one point to remove duplicate singleton relations (i.e. relations with exactly one clause) where the rules do not use constraints, aggregators, etc. So, while it would work for the disconnected case above, it wouldn't work for the aggregator rules you showed. It would also not work for these rules:
```
disconnected0() :- a(Z), b(Z), Z != 1.
disconnected1() :- a(Z), b(Z), Z != 1.
 d(X) :- c(X,Y), a(Y), disconnected0().
 d(X) :- c(X,Y), a(X), disconnected1().
```
because the disconnected rules would now also have a constraint, which is not supported.

One way to handle this problem is therefore to extend the check for bijective equivalence in `MinimiseProgramTransformer`  to cater for aggregators and constraints.

**Comment by azreika** [comment]
The program minimiser has been extended to support constraints and negations in PR #1516. Further extending it to cover clauses with aggregators should solve this issue. This can be done by extending the normalisation stage in the program-minimising transformation to also normalise aggregators, though care has to be taken to make sure aggregator bodies are distinguished from each other as well as from the outer scope.

**Comment by azreika** [comment]
Resolved by PR #1577.
